internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -0,0 +1,270 @@
+	if dhtm.checkForTimeoutsTimer != nil {
+		dhtm.checkForTimeoutsTimer.Stop()
+	}
+}
+
+// onStarted is called when the dontHaveTimeoutMgr starts.
+// It monitors for the context being cancelled.
+func (dhtm *dontHaveTimeoutMgr) onStarted() {
+	<-dhtm.ctx.Done()
+	dhtm.onShutdown()
+}
+
+// Start the dontHaveTimeoutMgr. This method is idempotent
+func (dhtm *dontHaveTimeoutMgr) Start() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Make sure the dont have timeout manager hasn't already been started
+	if dhtm.started {
+		return
+	}
+	dhtm.started = true
+
+	go dhtm.onStarted()
+
+	// If we already have a measure of latency to the peer, use it to
+	// calculate a reasonable timeout
+	latency := dhtm.peerConn.Latency()
+	if latency.Nanoseconds() > 0 {
+		dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+		return
+	}
+
+	// Otherwise measure latency by pinging the peer
+	go dhtm.measureLatency()
+}
+
+// measureLatency measures the latency to the peer by pinging it
+func (dhtm *dontHaveTimeoutMgr) measureLatency() {
+	// Wait up to defaultTimeout for a response to the ping
+	ctx, cancel := context.WithTimeout(dhtm.ctx, dhtm.defaultTimeout)
+	defer cancel()
+
+	// Ping the peer
+	res := dhtm.peerConn.Ping(ctx)
+	if res.Error != nil {
+		// If there was an error, we'll just leave the timeout as
+		// defaultTimeout
+		return
+	}
+
+	// Get the average latency to the peer
+	latency := dhtm.peerConn.Latency()
+
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Calculate a reasonable timeout based on latency
+	dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+
+	// Check if after changing the timeout there are any pending wants that are
+	// now over the timeout
+	go dhtm.checkForTimeouts()
+}
+
+// checkForTimeouts checks pending wants to see if any are over the timeout
+func (dhtm *dontHaveTimeoutMgr) checkForTimeouts() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	if len(dhtm.pending) == 0 {
+		return
+	}
+
+	// Figure out which of the blocks that were wanted were not received
+	// within the timeout
+	pending := make([]cid.Cid, 0, len(dhtm.pending))
+	var oldest time.Time
+	for c, started := range dhtm.pending {
+		if time.Since(started) > dhtm.timeout {
+			pending = append(pending, c)
+			delete(dhtm.pending, c)
+		} else if oldest.IsZero() || started.Before(oldest) {
+			oldest = started
+		}
+	}
+
+	// Fire the timeout event for the pending wants
+	if len(pending) > 0 {
+		go dhtm.fireTimeout(pending)
+	}
+
+	if len(dhtm.pending) == 0 {
+		return
+	}
+
+	// Make sure things are still running
+	select {
+	case <-dhtm.ctx.Done():
+		return
+	default:
+	}
+
+	// If there's an existing scheduled check, stop it
+	if dhtm.checkForTimeoutsTimer != nil {
+		dhtm.checkForTimeoutsTimer.Stop()
+	}
+
+	// Schedule the next check for the moment when the oldest pending want will
+	// timeout
+	until := time.Until(oldest.Add(dhtm.timeout))
+	dhtm.checkForTimeoutsTimer = time.AfterFunc(until, dhtm.checkForTimeouts)
```

Comment:
We don't need to repeatedly create new AfterFuncs. We can create one and then `Reset` the timeout to restart it.